gen/internals.py

```diff
@@ -0,0 +1,464 @@
+            kwargs[parameter] = self._get(parameter)
+
+        try:
+            value = setter.calc(**kwargs)
+        except AssertionError as ex:
+            self._errors[name] = ex.args[0]
+            raise CalculatorError("assertion while calc")
+
+        if name in self._validate_by_arg:
+            try:
+                self._validate_by_arg[name](value)
+            except AssertionError as ex:
+                self._errors[name] = ex.args[0]
+                raise CalculatorError("assertion while validate")
+
+        return value
+
+    def _get(self, name):
+        if name in self._arguments:
+            if self._arguments[name] is None:
+                raise CalculatorError("No way to set", [name])
+            return self._arguments[name]
+
+        # Detect cycles by checking if we're in the middle of calculating the
+        # argument being asked for
+        if name in self.__in_progress:
+            raise CalculatorError("Internal error. cycle detected. re-encountered {}".format(name))
+
+        self.__in_progress.add(name)
+        try:
+            self._arguments[name] = self._calculate_argument(name)
+            return self._arguments[name]
+        except CalculatorError as ex:
+            self._arguments[name] = None
+            raise CalculatorError(ex.message, ex.chain + ['while calculating {}'.format(name)]) from ex
+        except:
+            self._arguments[name] = None
+            raise
+        finally:
+            self.__in_progress.remove(name)
+
+    def _calculate_target(self, target):
+        def evaluate_var(name):
+            try:
+                self._get(name)
+            except CalculatorError as ex:
+                log.debug("Error calculating %s: %s. Chain: %s", name, ex.message, ex.chain)
+
+        for name in target.variables:
+            evaluate_var(name)
+
+        for name, sub_scope in target.sub_scopes.items():
+            if name not in self._arguments:
+                evaluate_var(name)
+
+            # If the internal arg is None, there was an error, don't check if it
+            # is a legal choice.
+            if self._arguments[name] is None:
+                continue
+
+            choice = self._get(name)
+
+            if choice not in sub_scope.cases:
+                self._errors[name] = "Invalid choice {}. Must choose one of {}".format(
+                    choice, ", ".join(sorted(sub_scope.keys())))
+                continue
+
+            self._calculate_target(sub_scope.cases[choice])
+
+        # Perform all multi-argument validations
+        for parameter_set, validate_fn in self._multi_arg_validate.items():
+            # Build up argument map for validate function. If any arguments are
+            # unset then skip this validate function.
+            kwargs = dict()
+            skip = False
+            for parameter in parameter_set:
+                if parameter not in self._arguments or self._arguments[parameter] is None:
```

Comment:
I would parenthesize this in order to prevent any tricky operator precedence rule that can go un-noticed.